### PR TITLE
make service alias unique

### PR DESCRIPTION
**Pull request by ilgooz**
a part of https://github.com/mesg-foundation/core/issues/618

### CODE REVIEW

File: database/service_db.go [hunk 1]
```diff
@@ -169,16 +169,36 @@ func (d *LevelDBServiceDB) Save(s *service.Service) error {
...
+	tx, err := d.db.OpenTransaction()
+	if err != nil {
+		return err
+	}
+
+	// check if there is a service with the same alias.
+	if _, err := d.get(tx, s.Alias); err == nil {
+		tx.Discard()
+		return &ErrSameAlias{s.Alias}
+	} else if _, ok := err.(*ErrNotFound); !ok {
```
Comment (NicolasMahe) [inline database/service_db.go]:
```suggestion
	} else if !IsErrNotFound(err) {
```

File: database/service_db_test.go [hunk 2]
```diff
@@ -48,6 +48,21 @@ func TestServiceDBSave(t *testing.T) {
...
+func TestServiceDBSaveWithSameAlias(t *testing.T) {
+	db, closer := openServiceDB(t)
+	defer closer()
+
+	s := &service.Service{ID: "01", Alias: "a", Name: "test-service"}
+	require.NoError(t, db.Save(s))
+	s = &service.Service{ID: "02", Alias: "a", Name: "test-service"}
+	require.Equal(t, &ErrSameAlias{"a"}, db.Save(s))
+
+	_, err := db.Get("1")
```
Comment (krhubert) [inline database/service_db_test.go]: Why to get `"1"` and `"2"` (as they are unrelated to this test), probably you wanted `"01"` and `"02"`

**Comment by ilgooz** [comment]
I'm freezing this PR for now. [Discussion](https://forum.mesg.com/t/brainstorming-services-alias-sid-versions-and-data-volumes/129/6) needs to be finalized first.

**Comment by NicolasMahe** [comment]
Anthony suggest to put the logic if the service alias already exist in the `api.deploy` function.
In this function, we have access to the service def, and the database. We also have a 2 way stream, so we can send the replace confirmation to the cli and wait for confirmation.
The deploy should not rely on database error but should call the database.get function to check if the alias already exist

**Comment by NicolasMahe** [comment]
This PR is only about adding a new behavior to the database package.
When saving a service with an alias that already exist in the db, existing service with same alias get deleted.
Result: only 1 service per alias can exist in the db

File: database/service_db_test.go [hunk 3]
```diff
@@ -30,13 +30,28 @@ func TestServiceDBSave(t *testing.T) {
...
+	s2 := &service.Service{ID: "01", Alias: "1", Name: "test-service"}
+	require.NoError(t, db.Save(s2))
+	_, err := db.Get(s1.ID)
+	require.IsType(t, &ErrNotFound{}, err)
+
+	// different id, different alias. should not replace anything
+	s3 := &service.Service{ID: "02", Alias: "2", Name: "test-service"}
+	require.NoError(t, db.Save(s3))
+	ss, _ = db.All()
+	require.Len(t, ss, 2)
```
Comment (ilgooz) [inline database/service_db_test.go]: it'd be better also to compare received services here with Equal so we can make sure that we really overwrite the service with new one(s2) correctly.

**Comment by ilgooz** [comment]
LGTM

**Pull request merged**